File: Solution/src/com/solution/dynamicP/KnapSack01.java
package com.solution.dynamicP;

import java.util.Scanner;

import com.solution.main.ReadInput;

public class KnapSack01 {

	public void run() {

		Scanner sc = ReadInput.getScanner("knapsack");
		sc.useDelimiter("\n");
		int size = sc.nextInt();
		int rw = sc.nextInt();

		String[] weight = sc.next().split(" ");
		String[] val = sc.next().split(" ");

		int[] w = new int[size];
		int[] v = new int[size];

		for (int i = 0; i < size; i++) {
			w[i] = Integer.parseInt(weight[i]);
			v[i] = Integer.parseInt(val[i]);
		}

		System.out.println("Recursion: " + doRunRec(w, v, rw, 0, 0));
		System.out.println("DP: " + doRunDP(w, v, rw));

	}

	/**
	 * 
	 * @param weight
	 *            array
	 * @param val
	 *            array
	 * @param w
	 *            required weight
	 * @param v
	 *            required value
	 * @param itemNumber
	 *            current Item
	 */
	private int doRunRec(int[] weight, int[] val, int w, int v, int itemNumber) {
		if (w == 0 || itemNumber == weight.length) {
			return 0;
		}

		if (w < weight[itemNumber]) {
			return doRunRec(weight, val, w, v, itemNumber + 1);
		}

		int lmax = val[itemNumber] + doRunRec(weight, val, w - weight[itemNumber], v + val[itemNumber], itemNumber + 1);
		int rmax = doRunRec(weight, val, w, v, itemNumber + 1);

		return Math.max(lmax, rmax);
	}

	private int doRunDP(int[] weight, int[] val, int rw) {

		int[][] T = new int[weight.length][rw + 1];

		for (int i = 0; i < weight.length; i++) {
			T[i][0] = 0;
			if (weight[i] < i) {
				T[0][i] = val[i];
			} else {
				T[0][i] = 0;
			}

		}

		for (int i = 1; i < weight.length; i++) {
			for (int j = 1; j < T[0].length; j++) {

				if (j < weight[i - 1]) {
					T[i][j] = T[i - 1][j]; // 'i-1' because we are 1 index ahead
											// of weight array.
				} else {
					T[i][j] = Math.max(val[i - 1] + T[i - 1][j - weight[i - 1]], T[i - 1][j]);
				}
			}
		}

		return T[weight.length - 1][rw];
	}

}

File: src/main/kotlin/com/cssolutions/dynamicProgramming/KnapSack01.kt
package com.cssolutions.dynamicProgramming

class KnapSack01 {

    fun knapsack01Recursion(weightArray: IntArray, profitArray: IntArray, requiredWeight: Int, profitSoFar: Int, itemNumber: Int): Int {
        if (requiredWeight == 0 || itemNumber == weightArray.size) {
            return 0
        }
        if (requiredWeight < weightArray[itemNumber]) {
            return knapsack01Recursion(weightArray, profitArray, requiredWeight, profitSoFar, itemNumber + 1)
        }
        val lmax = profitArray[itemNumber] + knapsack01Recursion(weightArray, profitArray, requiredWeight - weightArray[itemNumber], profitSoFar + profitArray[itemNumber], itemNumber + 1)
        val rmax = knapsack01Recursion(weightArray, profitArray, requiredWeight, profitSoFar, itemNumber + 1)
        return Math.max(lmax, rmax)
    }

    fun knapsack01DP(weightArray: IntArray, profitArray: IntArray, requiredWeight: Int): Int {
        val dpArray = Array(weightArray.size + 1) { IntArray(requiredWeight + 1) }
        for (i in weightArray.indices) {
            dpArray[i][0] = 0
            if (weightArray[i] < i) {
                dpArray[0][i] = profitArray[i]
            } else {
                dpArray[0][i] = 0
            }
        }
        for (i in 1 until dpArray.size) {
            for (j in 1 until dpArray[0].size) {
                if (j < weightArray[i - 1]) {
                    dpArray[i][j] = dpArray[i - 1][j] // 'i-1' because we are 1 index ahead
                    // of weight array.
                } else {
                    dpArray[i][j] = Math.max(profitArray[i - 1] + dpArray[i - 1][j - weightArray[i - 1]], dpArray[i - 1][j])
                }
            }
        }
        return dpArray[weightArray.size][requiredWeight]
    }
}